File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_py/code_output/AssessmentSystem.py
import json


class AssessmentSystem:
    def __init__(self):
        self.students = {}

    def addStudent(self, name, grade, major):
        self.students[name] = self.Student(name, grade, major)

    def addCourseScore(self, name, course, score):
        if name in self.students:
            self.students[name].addCourseScore(course, score)

    def getGPA(self, name):
        if name in self.students:
            return self.students[name].calculateGPA()
        return None

    def getAllStudentsWithFailCourse(self):
        failingStudents = []
        for student in self.students.values():
            if student.hasFailingCourse():
                failingStudents.append(student.getName())
        return failingStudents

    def getCourseAverage(self, course):
        totalScore = 0
        count = 0
        for student in self.students.values():
            score = student.getCourseScore(course)
            if score is not None:
                totalScore += score
                count += 1
        return totalScore / count if count > 0 else None

    def getTopStudent(self):
        topStudent = None
        topGPA = 0.0
        for student in self.students.values():
            gpa = student.calculateGPA()
            if gpa is not None and gpa > topGPA:
                topGPA = gpa
                topStudent = student.getName()
        return topStudent

    class Student:
        def __init__(self, name, grade, major):
            self.name = name
            self.grade = grade
            self.major = major
            self.courses = {}

        def getName(self):
            return self.name

        def addCourseScore(self, course, score):
            self.courses[course] = score

        def calculateGPA(self):
            if not self.courses:
                return None
            totalScore = sum(self.courses.values())
            return totalScore / len(self.courses)

        def hasFailingCourse(self):
            for score in self.courses.values():
                if score < 60:
                    return True
            return False

        def getCourseScore(self, course):
            return self.courses.get(course)

        def __eq__(self, other):
            if self is other:
                return True
            if other is None or type(self) != type(other):
                return False
            return (self.grade == other.grade and
                    self.name == other.name and
                    self.major == other.major and
                    self.courses == other.courses)

        def __hash__(self):
            return hash((self.name, self.grade, self.major, json.dumps(self.courses, sort_keys=True)))


if __name__ == "__main__":
    system = AssessmentSystem()
    system.addStudent("student 1", 3, "SE")
    system.addStudent("student 2", 2, "SE")
    system.addCourseScore("student 1", "course 1", 86)
    system.addCourseScore("student 2", "course 1", 59)
    system.addCourseScore("student 1", "course 2", 78)
    system.addCourseScore("student 2", "course 2", 90)

    print(system.getAllStudentsWithFailCourse())
    print(system.getCourseAverage("course 1"))
    print(system.getCourseAverage("course 2"))
    print(system.getGPA("student 1"))
    print(system.getGPA("student 2"))
    print(system.getTopStudent())

File: data/derived_cot/rq3_prompting/results/translation/glm-5.1/java_to_cpp/code_output/AssessmentSystem.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <optional>

class Student {
private:
    std::string name;
    int grade;
    std::string major;
    std::unordered_map<std::string, int> courses;

public:
    Student(std::string name, int grade, std::string major)
        : name(std::move(name)), grade(grade), major(std::move(major)) {}

    std::string getName() const { return name; }

    void addCourseScore(const std::string& course, int score) {
        courses[course] = score;
    }

    std::optional<double> calculateGPA() const {
        if (courses.empty()) {
            return std::nullopt;
        }
        int totalScore = 0;
        for (const auto& [c, score] : courses) {
            totalScore += score;
        }
        return static_cast<double>(totalScore) / courses.size();
    }

    bool hasFailingCourse() const {
        for (const auto& [c, score] : courses) {
            if (score < 60) {
                return true;
            }
        }
        return false;
    }

    std::optional<int> getCourseScore(const std::string& course) const {
        auto it = courses.find(course);
        if (it != courses.end()) {
            return it->second;
        }
        return std::nullopt;
    }
};

class AssessmentSystem {
    std::unordered_map<std::string, Student> students;

public:
    AssessmentSystem() = default;

    void addStudent(const std::string& name, int grade, const std::string& major) {
        students.emplace(name, Student(name, grade, major));
    }

    void addCourseScore(const std::string& name, const std::string& course, int score) {
        auto it = students.find(name);
        if (it != students.end()) {
            it->second.addCourseScore(course, score);
        }
    }

    std::optional<double> getGPA(const std::string& name) {
        auto it = students.find(name);
        if (it != students.end()) {
            return it->second.calculateGPA();
        }
        return std::nullopt;
    }

    std::vector<std::string> getAllStudentsWithFailCourse() {
        std::vector<std::string> failingStudents;
        for (auto& [n, student] : students) {
            if (student.hasFailingCourse()) {
                failingStudents.push_back(student.getName());
            }
        }
        return failingStudents;
    }

    std::optional<double> getCourseAverage(const std::string& course) {
        int totalScore = 0;
        int count = 0;
        for (auto& [n, student] : students) {
            auto score = student.getCourseScore(course);
            if (score.has_value()) {
                totalScore += score.value();
                count++;
            }
        }
        return count > 0 ? std::optional<double>(static_cast<double>(totalScore) / count) : std::nullopt;
    }

    std::string getTopStudent() {
        std::string topStudent;
        double topGPA = 0;
        for (auto& [n, student] : students) {
            auto gpa = student.calculateGPA();
            if (gpa.has_value() && gpa.value() > topGPA) {
                topGPA = gpa.value();
                topStudent = student.getName();
            }
        }
        return topStudent;
    }
};

static void printJavaDouble(double d) {
    long long i = static_cast<long long>(d);
    if (d == static_cast<double>(i)) {
        std::cout << i << ".0" << std::endl;
    } else {
        std::cout << d << std::endl;
    }
}

int main() {
    AssessmentSystem system;
    system.addStudent("student 1", 3, "SE");
    system.addStudent("student 2", 2, "SE");
    system.addCourseScore("student 1", "course 1", 86);
    system.addCourseScore("student 2", "course 1", 59);
    system.addCourseScore("student 1", "course 2", 78);
    system.addCourseScore("student 2", "course 2", 90);

    auto failing = system.getAllStudentsWithFailCourse();
    std::cout << "[";
    for (size_t i = 0; i < failing.size(); i++) {
        if (i > 0) std::cout << ", ";
        std::cout << failing[i];
    }
    std::cout << "]" << std::endl;

    auto avg1 = system.getCourseAverage("course 1");
    if (avg1.has_value()) printJavaDouble(avg1.value());
    else std::cout << "null" << std::endl;

    auto avg2 = system.getCourseAverage("course 2");
    if (avg2.has_value()) printJavaDouble(avg2.value());
    else std::cout << "null" << std::endl;

    auto gpa1 = system.getGPA("student 1");
    if (gpa1.has_value()) printJavaDouble(gpa1.value());
    else std::cout << "null" << std::endl;

    auto gpa2 = system.getGPA("student 2");
    if (gpa2.has_value()) printJavaDouble(gpa2.value());
    else std::cout << "null" << std::endl;

    std::string top = system.getTopStudent();
    if (top.empty()) std::cout << "null" << std::endl;
    else std::cout << top << std::endl;

    return 0;
}